### PR TITLE
Overlord - First Submission

### PR DESCRIPTION
**Pull request by oharace**
Ready for review.

To add custom activate/deactivate codes, modify line #62 in overlord.rb.
I focused primarily on testing and functionality, so the aesthetics are neither good nor fancy. I can add those later if desired or if time allows.

### CODE REVIEW

File: overlord/public/scripts/overlord.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
+"use strict";
+
+var deactivated = true;
+
+function submitCode() {
```
Comment (oharace) [inline overlord/public/scripts/overlord.js]: It is called from the form in the index.haml file.

File: overlord/public/scripts/overlord.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  document.getElementById("bombState").innerHTML = bombState;
+  document.getElementById("formLabel").innerHTML = deactivateLabel;
+  if (jsonObj.attempts_remaining <= 0) {
+    explodedUpdates();
+  } else if (jsonObj.attempts_remaining < 3) {
+    var attemptsRemaining = "(" + jsonObj.attempts_remaining;
+    attemptsRemaining = attemptsRemaining + " attempts remaining!)";
+    document.getElementById("AttemptsRemaining").innerHTML = attemptsRemaining;
+  }
+  timer(jsonObj.timer_end);
```
Comment (oharace) [inline overlord/public/scripts/overlord.js]: timer_end is from the json string that is returned by the ajax call. It is not a variable name, so it should be fine.

File: overlord/public/scripts/overlord.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
+var deactivated = true;
+
+function submitCode() {
```
Comment (oharace) [inline overlord/public/scripts/overlord.js]: It is called from the form in the index.haml file

File: overlord/features/manage_bomb.feature [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    Given the bomb is deactivated
+    When I enter an incorrect activation code
+    Then the bomb should be deactivated
+    And the timer should be stopped
+
+  @javascript
+  Scenario: Enter Incorrect Deactivation Code 3 Times
+    Given the bomb is activated
+    When I enter an incorrect deactivation code
+    And I enter an incorrect deactivation code again
```
Comment (magicalbanana) [inline overlord/features/manage_bomb.feature]: I want to see this as an outline or the very least, simplified.

File: overlord/features/manage_bomb.feature [hunk 5]
```diff
@@ -0,0 +1,60 @@
+Feature: Manage Bomb
+  In order to create a bomb
+  As a genius contractor
+  I want to create and manage the bomb
+
+  Background:
+    Given I am on the page
```
Comment (magicalbanana) [inline overlord/features/manage_bomb.feature]: Be more expressive, page of what? When a stake holder reads this test, they will most likely as you the same thing as I am.

File: overlord/spec/bomb_spec.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
+require_relative '../bomb.rb'
```
Comment (magicalbanana) [inline overlord/spec/bomb_spec.rb]: You can configure all that you need for your test via the spec_helper file.

But this is ok as well, but ideally when building a test suite you want a global configuration for your files.

File: overlord/views/index.haml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+!!!
+%html
+  %head
+    %title Overlord!
+    %link{:rel => :stylesheet, :type => :"text/css", :href => "styles/styles.css"}
+    %script{:src => "//code.jquery.com/jquery-1.11.3.min.js"}
```
Comment (magicalbanana) [inline overlord/views/index.haml]: Good job on sourcing this than actually adding it to your project app.

File: overlord/spec/bomb_spec.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+describe Bomb do
+  before :each do
+    @bomb = Bomb.new
+  end
+
+  it "is deactivated" do
```
Comment (magicalbanana) [inline overlord/spec/bomb_spec.rb]: I want you to look at betterspecs.org.

Be familiar with contexts and using describe more expressively.

File: overlord/bomb.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  def explode
+    @state = "Exploded"
+  end
+
+  def start_timer
+    @timer_end = (Time.now + TWO_MINUTES).to_i
+  end
+
+  def correct_activation_code?(code)
+    code == @default_activation_code || code == @custom_activation_code
```
Comment (paulghaddad) [inline overlord/bomb.rb]: Consider extracting each side of this `II` operator as a method.

File: overlord/features/manage_bomb.feature [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    And the bomb should be exploded
+    And the buttons should be disabled
+    And the timer should be stopped
+
+  @javascript
+  Scenario: The bomb timer reaches 00:00
+    Given the bomb is activated
+    When the bomb timer ends
+    Then the bomb should explode
+    And the bomb should be exploded
```
Comment (paulghaddad) [inline overlord/features/manage_bomb.feature]: Try writing your step definitions at a higher level of abstraction. Your `And` statements here are too fine grained. See this article as reference:

http://robots.thoughtbot.com/acceptance-tests-at-a-single-level-of-abstraction

File: overlord/features/step_definitions/bomb_steps.rb [hunk 11]
```diff
@@ -0,0 +1,100 @@
+Given(/^I am on the page$/) do
+  visit '/'
+  sleep(0.5)
+  @bomb = Bomb.new("2345", "5432")
+  # save_and_open_page
```
Comment (paulghaddad) [inline overlord/features/step_definitions/bomb_steps.rb]: Comment

File: overlord/features/step_definitions/bomb_steps.rb [hunk 12]
```diff
@@ -0,0 +1,100 @@
+Given(/^I am on the page$/) do
+  visit '/'
+  sleep(0.5)
+  @bomb = Bomb.new("2345", "5432")
```
Comment (paulghaddad) [inline overlord/features/step_definitions/bomb_steps.rb]: Why are you creating a bomb here? Shouldn't this be done through the UI?

File: overlord/features/support/env.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
+require_relative './../../bomb.rb'
```
Comment (paulghaddad) [inline overlord/features/support/env.rb]: This is a test smell. You shouldn't need to include the `bomb.rb` file for feature tests. The work should be accomplished through the UI.

**Comment by paulghaddad** [comment]
@oharace Take a look at our comments and please submit a second iteration. For this iteration, focus on your feature tests. You should perform your steps through the UI; also focus on ensuring your scenarios are clear and at a high level of abstraction. I'd be glad to discuss this in person.

**Pull request closed without merging**